### PR TITLE
Add needs-tws plugin code.

### PR DESCRIPTION
**Pull request by Ressetkk**
The plugin works as microservice. It checks if the Markdown files in the PR have changed and if so the PR gets flagged with `needs-tws-review` Label. It removes the label once one of the Technical Writers group, defined in OWNERS_ALIASES `technical-writers` alias approves the changes.

If Pull Request is opened/reopened/updated the plugin will fetch current changes in the PR and add or remove label if any of the Markdown files has been updated.
If Pull Request has been unlabelled the plugin makes the following checks:
* If the bot removed label - skip. 
* If non-collaborator removed label - re-add label and post comment that only collaborators with write access can remove label.
* If a collaborator with write access removes label - treat it as manual dismissal and do not add label.

If a PR is reviewed by someone:
* If a PR is reviewed by person who is not in `technical-writers` group - skip checking, as this review is for something else.
* If a PR is approved by a technical writer - (if needed) remove label.
* If technical write requested changed - (if needed) add label.
Authors and non-collaborators cannot review PR.

/hold

#4520

### CODE REVIEW

**Comment by Ressetkk** [comment]
Still in progress of adding tests.

/hold

**Comment by dekiel** [comment]
@Ressetkk do we still need this plugin? As you mention we will add tws as approvers for doc directories only right?

**Comment by Ressetkk** [comment]
That was one of the idea, however @kyma-project/technical-writers still wish to review and maintain all markdown files in the repositories so unless prow gets updated the plug-in will still be needed. It requires some real life testing period though, despite working tests

File: development/external-plugins/needs-tws/plugin.go
Comment (dekiel) [inline development/external-plugins/needs-tws/plugin.go:194]: Why we allow collaborators to bypass documentation review?

File: development/external-plugins/needs-tws/plugin.go
Comment (dekiel) [inline development/external-plugins/needs-tws/plugin.go:81]: Does it make sens to have two HandlePR and handlePullRequest methods? Why we need this separation? If this methods are doing something different the names should indicate that. If they do the same logic, why this is not one method.

File: development/external-plugins/needs-tws/plugin.go
Comment (dekiel) [inline development/external-plugins/needs-tws/plugin.go:103]: Same here. We have HandlePRReview then handlePullRequestReview and finally handle. This additional methods are not used in other places. They are not meant to be used outside this tool.

File: development/external-plugins/needs-tws/plugin.go
Comment (dekiel) [inline development/external-plugins/needs-tws/plugin.go:244]: This should return some error with description. At present in logs we will get `l.Errorw("Failed to handle Pull Request Review event", "error", err.Error())` where `err.Error()` will be simple a nil. That will be confusing when troubleshooting.

File: development/external-plugins/needs-tws/plugin.go
Comment (dekiel) [inline development/external-plugins/needs-tws/plugin.go:226]: This should return some error with description. At present in logs we will get l.Errorw("Failed to handle Pull Request Review event", "error", err.Error()) where err.Error() will be simple a nil. That will be confusing when troubleshooting.

File: development/external-plugins/needs-tws/plugin.go
Comment (dekiel) [inline development/external-plugins/needs-tws/plugin.go:278]: This should return some error with description. At present in logs we will get l.Errorw("Failed to handle Pull Request Review event", "error", err.Error()) where err.Error() will be simple a nil. That will be confusing when troubleshooting.

**Review by dekiel — CHANGES_REQUESTED** [review]
Please align external-plugins/needs-tws directory structure to match other plugins.
Please change plugin k8s resources yaml file name to indicate this is an external plugin. Same  as other external-plugin.
Please add some comments to the code to explain what is it's purpose or what is a context. Especialy  PluginBackend.handle method contains multiple conditions and assumptions which are not easy to follow.

File: development/external-plugins/needs-tws/plugin.go
Comment (Ressetkk) [reply development/external-plugins/needs-tws/plugin.go:183]: Renamed to `twsLabel`. It's hardcoded right now because in the future this variable should be taken from plugin config file, as the TODO says.

File: development/external-plugins/needs-tws/plugin.go
Comment (Ressetkk) [reply development/external-plugins/needs-tws/plugin.go:81]: Renamed to `PullRequestHandler`.

File: development/external-plugins/needs-tws/plugin.go
Comment (Ressetkk) [reply development/external-plugins/needs-tws/plugin.go]: Renamed to `PullRequestReviewHandler`. This function is meant to use with external plugin framework as a wrapper for previous implementation of a plugin.

File: development/external-plugins/needs-tws/plugin.go
Comment (Ressetkk) [reply development/external-plugins/needs-tws/plugin.go:248]: This event should be ignore. If it's not meant to be ignored the function will return proper `err`, and will be seen on logs. Added explanation comment in the code.

File: development/external-plugins/needs-tws/plugin.go
Comment (Ressetkk) [reply development/external-plugins/needs-tws/plugin.go:229]: This event should be ignore. If it's not meant to be ignored the function will return proper `err`, and will be seen on logs. Added explanation comment in the code.

File: development/external-plugins/needs-tws/plugin.go
Comment (Ressetkk) [reply development/external-plugins/needs-tws/plugin.go:285]: This event should be ignore. If it's not meant to be ignored the function will return proper `err`, and will be seen on logs. Added explanation comment in the code.

File: development/external-plugins/needs-tws/plugin.go
Comment (Ressetkk) [reply development/external-plugins/needs-tws/plugin.go:197]: The plugin is meant to work when all collaborators and team members have their write access revoked, so the Prow is the one managing labels and Tide is merging PRs based on the state of labels. Note that this check will only work when the PR is **unlabeled*, that means the person is removing the label. This restricts to the people with the write access only.

If the people will need write access to the repo then for blocking purposes this plugin should not be used and the GitHub's `CODEOWNERS` should be used instead.

File: development/external-plugins/needs-tws/plugin.go
Comment (Ressetkk) [reply development/external-plugins/needs-tws/plugin.go]: It's meant to work to opened PRs which are ready for review, that's why Drafts are filtered.

File: development/external-plugins/needs-tws/plugin.go
Comment (Ressetkk) [reply development/external-plugins/needs-tws/plugin.go]: There is no need to handle Closed PRs at all. This line ensure the event is discarded before doing anything

File: development/external-plugins/needs-tws/plugin.go
Comment (Ressetkk) [reply development/external-plugins/needs-tws/plugin.go]: I'll remove the check ultimately. On second thought having this label added to the Drafts is a better approach for informing a person what is missing.

**Review by dekiel — APPROVED** [review]

**Pull request merged**
